lib/Parse/SyntaxParseActions.cpp

```diff
@@ -0,0 +1,170 @@
+//===--- SyntaxParseActions.cpp -------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Parse/SyntaxParseActions.h"
+#include "swift/Parse/Token.h"
+#include "swift/Syntax/SyntaxKind.h"
+
+using namespace swift;
+
+struct DeferredTokenNode {
+  bool IsMissing;
+  tok TokenKind;
+  StringRef LeadingTrivia;
+  StringRef TrailingTrivia;
+  /// The range of the token including trivia.
+  CharSourceRange Range;
+
+  DeferredTokenNode(bool IsMissing, tok TokenKind, StringRef LeadingTrivia,
+                    StringRef TrailingTrivia, CharSourceRange Range)
+      : IsMissing(IsMissing), TokenKind(TokenKind),
+        LeadingTrivia(LeadingTrivia), TrailingTrivia(TrailingTrivia),
+        Range(Range) {}
+};
+
+struct DeferredLayoutNode {
+  syntax::SyntaxKind Kind;
+  ArrayRef<RecordedOrDeferredNode> Children;
+  unsigned Length;
+
+  DeferredLayoutNode(syntax::SyntaxKind Kind,
+                     ArrayRef<RecordedOrDeferredNode> Children, unsigned Length)
+      : Kind(Kind), Children(Children), Length(Length) {}
+};
+
+OpaqueSyntaxNode SyntaxParseActions::makeDeferredToken(tok tokenKind,
+                                                       StringRef leadingTrivia,
+                                                       StringRef trailingTrivia,
+                                                       CharSourceRange range,
+                                                       bool isMissing) {
+  return new (DeferredNodeAllocator) DeferredTokenNode(
+      isMissing, tokenKind, leadingTrivia, trailingTrivia, range);
+}
+
+OpaqueSyntaxNode SyntaxParseActions::makeDeferredLayout(
+    syntax::SyntaxKind k, bool isMissing,
+    const ArrayRef<RecordedOrDeferredNode> &children) {
+  assert(!isMissing && "Missing layout nodes not implemented yet");
+
+  // Compute the length of this node.
+  unsigned length = 0;
+  for (auto &child : children) {
+    switch (child.getKind()) {
+    case RecordedOrDeferredNode::Kind::Null:
+      break;
+    case RecordedOrDeferredNode::Kind::Recorded:
+      llvm_unreachable("Children of deferred nodes must also be deferred");
```

Comment:
Do we have any comments why deferred nodes can't contain recorded nodes? If not, could you add some?

Reply:
It’s because of the way we create the syntax tree. Once we put the `SyntaxParsingContext` in deferred node all (child) nodes are also deferred. If we chose to record child nodes of deferred nodes, we would also get into trouble should we decide to discard the deferred node, because now we’d also need to discard the recorded child node, which cannot be done. Please correct me if any of the assumptions I’m making here are wrong.

We don’t have a comment about this anywhere. I’ll add one with (more or less) the text I mentioned above.